src/v/transform/commit_batcher.cc

```diff
@@ -0,0 +1,205 @@
+            vlog(
+              tlog.warn,
+              "unable to determine key ({}) coordinator: {}",
+              key,
+              cluster::error_category().message(int(result.error())));
+            it = _unbatched.upper_bound(key);
+            continue;
+        }
+        was_all_failures = false;
+        model::partition_id coordinator = result.value();
+        auto entry = _unbatched.extract(key);
+        // It's possible that the value was removed while the request was being
+        // made, in that case just skip the update.
+        if (!entry) {
+            it = _unbatched.upper_bound(key);
+            continue;
+        }
+        _batched[coordinator].insert_or_assign(entry.key(), entry.mapped());
+        _coordinator_cache[key] = coordinator;
+        if (!_timer.armed()) {
+            _timer.arm(_commit_interval);
+        }
+        it = _unbatched.upper_bound(key);
+    }
+    co_return was_all_failures;
+}
+
+template<typename ClockType>
+ss::future<> commit_batcher<ClockType>::stop() {
+    _timer.cancel();
+    _as.request_abort();
+    _unbatched_cond_var.signal();
+    co_await _gate.close();
+    co_await flush();
+}
+
+template<typename ClockType>
+ss::future<> commit_batcher<ClockType>::wait_for_previous_flushes(
+  model::transform_offsets_key, ss::abort_source* as) {
+    try {
+        co_await ssx::sleep_abortable<ClockType>(_commit_interval, *as, _as);
+    } catch (const ss::sleep_aborted&) {
+        // do nothing as we're shutting down, callers will do the right thing
+    }
+}
+
+template<typename ClockType>
+void commit_batcher<ClockType>::preload(model::transform_offsets_key) {
+    // For now, don't do anything, we will resolve it first time they key needs
+    // to commit a batch. In the future we can preload this, mostly to surface
+    // any errors (like too many keys, etc) before reading and processing any
+    // batches.
+}
+
+template<typename ClockType>
+void commit_batcher<ClockType>::unload(model::transform_offsets_key key) {
+    // Erase the key from the coordinator cache so that we don't
+    _coordinator_cache.erase(key);
+    _unbatched.erase(key);
+    // We don't need to clear batched, we can allow it to be flushed async.
+}
+
+template<typename ClockType>
+ss::future<> commit_batcher<ClockType>::commit_offset(
+  model::transform_offsets_key k, model::transform_offsets_value v) {
+    auto _ = _gate.hold();
+    auto it = _coordinator_cache.find(k);
+    if (it == _coordinator_cache.end()) {
+        _unbatched.insert_or_assign(k, v);
+        _unbatched_cond_var.signal();
+    } else {
+        _batched[it->second].insert_or_assign(k, v);
+        if (!_timer.armed()) {
+            _timer.arm(_commit_interval);
+        }
+    }
+    return ss::now();
+}
+
+template<typename ClockType>
+ss::future<> commit_batcher<ClockType>::flush() {
+    absl::btree_map<model::partition_id, kv_map> batched;
+    _batched.swap(batched);
+    co_await ss::parallel_for_each(
```

Comment:
what's the expected parallelism here and should it be constrained?

Reply:
The expected parallelism is 1 for now (we only use a single partition). But I added an arbitrary limit of 10 to future proof a bit.